### ISSUE TITLE
Java constructor annotations on classes

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-compiler#2153]
I used to be able to do stuff like:

```ceylon
import javax.inject { inject }

shared service inject
class Application(Repository repository) { ... }
```

which was nice. Now I get an error on `inject`:

```
Annotated program element does not satisfy annotation constraints: ClassDeclaration is not assignable to FunctionOrValueDeclaration|ValueDeclaration|ConstructorDeclaration
```

**Comment by FroMage** [comment]
Yes, @gavinking said that for classes with default constructors, constructor annotations should go on explicit constructors only, so you have to refactor your class to make an explicit constructor. Lobby ceylon-spec for a change, sorry.

**Comment by jvasileff** [comment]
Ok, thanks for the info.

**Comment by gavinking** [comment]
@FroMage Huh?! This class does not have a default constructor, it has an initializer parameter list!

I strongly believe that this code should work!

**Comment by gavinking** [comment]
>  Lobby ceylon-spec for a change, sorry.

It is definitely not `ceylon-spec` that is preventing this code from compiling!

**Comment by FroMage** [comment]
Oh come on, I asked you just that last week and you said we should not allow it because it's ambiguous if the annotation can be placed on both the type and the constructor on which construct we should put it.

**Comment by gavinking** [comment]
Well I don't recall what I said, but this case is not ambiguous...

**Comment by FroMage** [comment]
I recall what you said, that we can't allow constructor annotations on classes because it's ambiguous.

**Comment by luolong** [comment]
@gavinking, I think what @FroMage is trying to say is that since `javax.inject.Inject` annotation has both class and constructor as it's targets, applying it to Ceylon class with initializer section is ambiguous, as in Ceylon, classes and default constructors are sort of merged.

In this case though, I suspect that we could try and make a safe guess at what the developer means when annotating ceylon classes.

If class has initialization parameters block, like in @jvasileff's example, the annotation only makes sense when applying it to the constructor.

When initialization arguments list is empty, the annotation should most likely be applied to the class.

Of course this only applies since we know what is the intended effect of `javax.inject.Inject` annotatioin. 

What if I have a following Java annotation:
```java
@Retention(RetentionPolicy.SOURCE)
@Target({ElementType.CONSTRUCTOR, ElementType.TYPE, ElementType.METHOD})
public @interface Foo {}
```
What should be the proper behavior in this case?

**Comment by jvasileff** [comment]
@luolong, @tombentley solved this problem for *Java* annotations in https://github.com/ceylon/ceylon-compiler/issues/1471, and my hope would be that some similar solution for *Ceylon* annotations would be possible with something like https://github.com/ceylon/ceylon-spec/issues/1205.

But I really think disallowing the initialization parameter style classes from participating *at all* in constructor annotations really devalues the feature, and leads to a bunch of ugly boilerplate with explicit constructors.

**Comment by gavinking** [comment]
> since `javax.inject.Inject` annotation has both class and constructor as it's targets

But it doesn't! `@Inject` can't be applied to a Java `class`!

**Comment by gavinking** [comment]
> But I really think disallowing the initialization parameter style classes from participating *at all* in constructor annotations really devalues the feature, and leads to a bunch of ugly boilerplate with explicit constructors.

I agree. And I don't recall precisely what I said to @FroMage the other day, but if I said something that implies we should not be able to apply `@Inject` to a Ceylon `class`, then I guess I was wrong.

I mean, sure, we have the workaround of `inject__CONSTRUCTOR` but that's just plain ugly for such a commonly-used thing as constructor injection.

**Comment by luolong** [comment]
> But it doesn't! `@Inject` can't be applied to a Java class!

You are right of course...
